File: android/app/src/main/java/com/github/ai/simplesplit/android/data/json/JsonSerializer.kt
package com.github.ai.simplesplit.android.data.json

import arrow.core.Either
import arrow.core.raise.either
import com.github.ai.simplesplit.android.model.exception.ParsingException
import com.google.gson.GsonBuilder
import com.google.gson.JsonSyntaxException

class JsonSerializer {

    val gson = GsonBuilder()
        .setPrettyPrinting()
        .create()

    inline fun <reified T> deserialize(text: String): Either<ParsingException, T> =
        either {
            try {
                gson.fromJson(text, T::class.java)
            } catch (exception: JsonSyntaxException) {
                raise(ParsingException(cause = exception))
            }
        }

    inline fun <reified T> serialize(data: T): String {
        return gson.toJson(data)
    }
}

File: backend/app/src/main/scala/com/github/ai/split/data/JsonSerializer.scala
package com.github.ai.split.data

import com.github.ai.split.entity.exception.{DomainError, JsonDeserializationError, ParsingError}
import com.github.ai.split.utils.some
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import zio.{IO, Task, ZIO}

class JsonSerializer {

  private val gson = GsonBuilder().setPrettyPrinting().create()

  def serialize(data: Any): String = {
    gson.toJson(data)
  }

  def deserializer[T](
    data: Task[String],
    typeOf: Class[T]
  ): IO[ParsingError, T] = {
    for {
      json <- data.mapError(error => ParsingError(message = "Unable to read json data"))
      result <- ZIO
        .attempt {
          gson.fromJson(json, TypeToken.get(typeOf))
        }
        .mapError(error => JsonDeserializationError(typeOf = typeOf, cause = error.some))
    } yield result
  }
}
